### ISSUE TITLE
Unable to run Sitespeed in parallel

### ISSUE THREAD

**Issue by amitmalhotra-cell** [sitespeedio/sitespeed.io#3326]
hi @soulgalore 
Problem Statement : Unable to run Sitespeed in parallel.

### Prerequiste:
#Run the sitespeed docker image with customized entrypoint "/usr/bin/bash"
$ docker run -it --name sitespeedcustom --entrypoint "/usr/bin/bash" sitespeedio/sitespeed.io

### Steps : 
1) ### Run multiple login script in parallel inside 1 container
create a shell script to run 2 commands in parallel inside a container as below 
`command1="/start.sh --outputFolder /sitespeed.io/result1 /sitespeed.io/login.js --multi --config /sitespeed.io/config.json --influxdb.tags \"market=$test_region,dateTime=$dateTime\" --browsertime.my.env $env --browsertime.my.usrNm ${!USERNAME} --browsertime.my.pswd ${!PASSWORD}"
		
command2="/start.sh --outputFolder /sitespeed.io/result2 /sitespeed.io/login.js --multi --config /sitespeed.io/config.json --influxdb.tags \"market=$test_region,dateTime=$dateTime\" --browsertime.my.env $env --browsertime.my.usrNm ${!USERNAME} --browsertime.my.pswd ${!PASSWORD}"

eval $command1 & eval $command2 &                                           //runs both login in parallel
wait < <(jobs -p)
        `

2) Run above shell script , but got error as 
### ERROR : Websocket is not open: readyState 2 (CLOSING)
![image](https://user-images.githubusercontent.com/63041681/112830736-926a0e80-90b0-11eb-9aaf-9ae60078ba89.png)

Question : I want to simulate 20000 users in parallel , but seems only 1 user login per container is possible. Please confirm if true
because if i run multiple containers in parallel , it works fine , but not practical to run 20000 containers . Please advise.

**Comment by soulgalore** [comment]
I would use a load testing tool to run 20000 requests in stead of sitespeed.io I think it makes more sense.

For the web socket: I didn't fully follow with the multiple login in the same container, what do want to accomplish?

**Comment by AJ84064** [comment]
hi @soulgalore ,
I agree , load testing tool like jmeter should be used but it only captures server side performance not client side as sitespeed does. 

**GOAL : Load testing for 20000 users login in parallel and capture sitespeed performance result.**

**I THINK THIS IS POSSIBLE WITH NEW AWS FEATURE : [AWS LAMBDA WITH CUSTOM (SITESPEED) IMAGE AS RUNTIME](https://docs.aws.amazon.com/lambda/latest/dg/images-create.html) . IF THIS IS SUCCESSFUL , SITESPEED CAN BE RUN FROM API HENCE WE CAN ACHIEVE OUR GOAL.**

**BUT AWS LAMBDA has it limitations :** 
1) The container image can only run on a read-only file system. Your function code can access a writable /tmp directory. If you are using an image (like sitespeed) that requires a writable directory outside of /tmp, configure it to write to a directory under the /tmp directory.
2) To use custom image(like sitespeed) as runtime for lambda . it needs to be made compatible with lambda (https://docs.aws.amazon.com/lambda/latest/dg/images-create.html#images-create-2)
3) Since its serverless platform , we can only use headless chrome and no xvfb.

I tried to create a custom sitespeed image compatible with lambda but was not working (but this is another long story).
**TO OVERCOME ABOVE LIMITATIONS :** 
**STEP1 :  We were able to run headless chrome in lambda using puppeteer :** 
main.js having login code looks like below : 

const chromium = require('chrome-aws-lambda');
async function handler(event, context, callback) {
if(!browser) {
      writelog('browser instance does not exist');
      // if browser is not available in the scope of lambda execution.
      // launch a new browser
      browser = await chromium.puppeteer.launch({
        args: chromium.args,
        defaultViewport: chromium.defaultViewport,
        executablePath: await chromium.executablePath,
        headless: chromium.headless,
        ignoreHTTPSErrors: true,
      });
      writelog('browser instance launched '+browser);
    } else {
      writelog('browser instance re-used '+browser);
    }
//login successfully
    await page.waitForSelector(usernameelement);
    await page.type(usernameelement, 'usernamehere');
    await page.waitForSelector(passwordelement);
    await page.type(passwordelement, passwordhere);    
    await page.waitForSelector(loginbutton);
    await page.click(loginbutton);

**STEP2 (Question) : HOW TO USE SITESPEED with above chrome browser and capture sitespeed result for login scenario. (like import only sitespeed ?)**

**Comment by amitmalhotra-cell** [comment]
hi @soulgalore ,
I agree , load testing tool like jmeter should be used but it only captures server side performance not client side as sitespeed does. 

**GOAL : Load testing for 20000 users login in parallel and capture sitespeed performance result.**

**I THINK THIS IS POSSIBLE WITH NEW AWS FEATURE : [AWS LAMBDA WITH CUSTOM (SITESPEED) IMAGE AS RUNTIME](https://docs.aws.amazon.com/lambda/latest/dg/images-create.html) . IF THIS IS SUCCESSFUL , SITESPEED CAN BE RUN FROM API HENCE WE CAN ACHIEVE OUR GOAL.**

**BUT AWS LAMBDA has it limitations :** 
1) The container image can only run on a read-only file system. Your function code can access a writable /tmp directory. If you are using an image (like sitespeed) that requires a writable directory outside of /tmp, configure it to write to a directory under the /tmp directory.
2) To use custom image(like sitespeed) as runtime for lambda . it needs to be made compatible with lambda (https://docs.aws.amazon.com/lambda/latest/dg/images-create.html#images-create-2)
3) Since its serverless platform , we can only use headless chrome and no xvfb.

I tried to create a custom sitespeed image compatible with lambda but was not working (but this is another long story).
**TO OVERCOME ABOVE LIMITATIONS :** 
**STEP1 :  We were able to run headless chrome in lambda using puppeteer :** 
main.js having login code looks like below : 

const chromium = require('chrome-aws-lambda');
async function handler(event, context, callback) {
if(!browser) {
      writelog('browser instance does not exist');
      // if browser is not available in the scope of lambda execution.
      // launch a new browser
      browser = await chromium.puppeteer.launch({
        args: chromium.args,
        defaultViewport: chromium.defaultViewport,
        executablePath: await chromium.executablePath,
        headless: chromium.headless,
        ignoreHTTPSErrors: true,
      });
      writelog('browser instance launched '+browser);
    } else {
      writelog('browser instance re-used '+browser);
    }
//login successfully
    await page.waitForSelector(usernameelement);
    await page.type(usernameelement, 'usernamehere');
    await page.waitForSelector(passwordelement);
    await page.type(passwordelement, passwordhere);    
    await page.waitForSelector(loginbutton);
    await page.click(loginbutton);

**STEP2 (Question) : HOW TO USE SITESPEED with above chrome browser and capture sitespeed result for login scenario. (like import only sitespeed ?)**
